lib/logstash/plugin_mixins/rabbitmq_connection.rb

```diff
@@ -0,0 +1,158 @@
+# encoding: utf-8
+require "logstash/outputs/base"
+require "logstash/namespace"
+require "march_hare"
+require "java"
+
+# Common functionality for the rabbitmq input/output
+module LogStash::PluginMixins::RabbitMQConnection
+  EXCHANGE_TYPES = ["fanout", "direct", "topic"]
+
+  HareInfo = Struct.new(:connection, :channel, :exchange)
+
+  def self.included(base)
+    base.extend(self)
+    base.setup_rabbitmq_connection_config
+  end
+
+  def setup_rabbitmq_connection_config
+    # RabbitMQ server address
+    config :host, :validate => :string, :required => true
+
+    # RabbitMQ port to connect on
+    config :port, :validate => :number, :default => 5672
+
+    # RabbitMQ username
+    config :user, :validate => :string, :default => "guest"
+
+    # RabbitMQ password
+    config :password, :validate => :password, :default => "guest"
+
+    # The vhost to use. If you don't know what this is, leave the default.
+    config :vhost, :validate => :string, :default => "/"
+
+    # Enable or disable SSL
+    config :ssl, :validate => :boolean, :default => false
+
+    # Validate SSL certificate
+    config :verify_ssl, :validate => :boolean, :default => false
+
+    # Enable or disable logging
+    config :debug, :validate => :boolean, :default => false, :deprecated => "Use the logstash --debug flag for this instead."
+
+    # Set this to automatically recover from a broken connection. You almost certainly don't want to override this!!!
+    config :automatic_recovery, :validate => :boolean, :default => true
+
+    # The exchange type (fanout, topic, direct)
+    config :exchange_type, :validate => EXCHANGE_TYPES, :required => true
+
+    # The name of the exchange
+    config :exchange, :validate => :string, :required => true
+
+    # Key to route to by default. Defaults to 'logstash'
+    #
+    # * Routing keys are ignored on fanout exchanges.
+    config :key, :validate => :string, :default => "logstash"
+
+    # Is this exchange durable? (aka; Should it survive a broker restart?)
+    config :durable, :validate => :boolean, :default => true
+
+    # Should RabbitMQ persist messages to disk?
+    config :persistent, :validate => :boolean, :default => true
+
+    # Time in seconds to wait before retrying a connection
+    config :connect_retry_interval, :validate => :number, :default => 1
+  end
+
+  def publish(event, message)
+    @hare_info.exchange.publish(message, :routing_key => event.sprintf(@key), :properties => { :persistent => @persistent })
+  rescue MarchHare::Exception, IOError, com.rabbitmq.client.AlreadyClosedException => e
+    return if terminating?
+
+    @logger.error("Error while publishing. Will retry.",
+                  :message => e.message,
+                  :exception => e.class,
+                  :backtrace => e.backtrace)
+
+    sleep_for_retry
+    retry
+  end
+
+  def conn_str
+    return "amqp://#{@user}@#{@host}:#{@port}#{@vhost}/#{@exchange_type}/#{@exchange}\##{@key}"
```

Comment:
Ask and ye shall receive! (fixed)

On Thu, Aug 20, 2015 at 5:20 PM, Guy Boertje <notifications@github.com>
wrote:

> In lib/logstash/plugin_mixins/rabbitmq_connection.rb
> <https://github.com/logstash-plugins/logstash-mixin-rabbitmq_connection/pull/1#discussion_r37587875>
> :
>
> > +  def publish(event, message)
> > +    @hare_info.exchange.publish(message, :routing_key => event.sprintf(@key), :properties => { :persistent => @persistent })
> > +  rescue MarchHare::Exception, IOError, com.rabbitmq.client.AlreadyClosedException => e
> > +    return if terminating?
> > +
> > +    @logger.error("Error while publishing. Will retry.",
> > +                  :message => e.message,
> > +                  :exception => e.class,
> > +                  :backtrace => e.backtrace)
> > +
> > +    sleep_for_retry
> > +    retry
> > +  end
> > +
> > +  def conn_str
> > +    return "amqp://#{@user}@#{@host}:#{@port}#{@vhost}/#{@exchange_type}/#{@exchange}\##{@key}"
>
> (pleading) explicit return here is v unnecessary
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/logstash-plugins/logstash-mixin-rabbitmq_connection/pull/1/files#r37587875>
> .
>